### PR TITLE
[Serve] refactor HTTPProxy into GenericProxy

### LINKED ISSUE #38193: [Serve] Productionize gRPC proxy work
### Description

We have explored all options and finalized interface for using the new gRPC proxy. Let's clean up the code and make sure all parts are tested. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Refactor `HTTPProxy` into `GenericProxy` so we can use in `gRPCProxy` later on. No new code, just reorganize methods and determine the ones that need to be implemented vs can be shared between the subclasses.

## Related issue number

First PR to close https://github.com/ray-project/ray/issues/38193

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -199,12 +199,20 @@ def match_route(
         return None
 
 
-class HTTPProxy:
-    """This class is meant to be instantiated and run by an ASGI HTTP server.
-
-    >>> import uvicorn
-    >>> controller_name = ... # doctest: +SKIP
-    >>> uvicorn.run(HTTPProxy(controller_name)) # doctest: +SKIP
+class GenericProxy:
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:202]: mark this as an abstract base class: https://docs.python.org/3/library/abc.html

and methods that need to be implement should be abstract methods

also, any method that needs to be implemented should not be underscore-prefixed

File: python/ray/serve/_private/http_proxy.py [hunk 2]
```diff
@@ -199,12 +199,20 @@ def match_route(
...
+    The proxy subclass need to implement the following methods:
+      - `proxy_name()`
+      - `_not_found()`
+      - `_draining_response()`
+      - `_timeout_response()`
+      - `_routes_response()`
+      - `_health_response()`
+      - `send_request_to_replica_unary()`
+      - `setup_request_context_and_handle()`
+      - `send_request_to_replica_streaming()`
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:215]: why do these need to be different across the impls?

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:215]: Bc how gRPC and ASGI handling request and response are very different. `_assign_request_with_timeout()` is still reused, but unwrapping receive object and construct response al need to be implemented differently in `send_request_to_replica_streaming()`

File: python/ray/serve/_private/http_proxy.py [hunk 3]
```diff
@@ -409,11 +410,12 @@ def _ongoing_requests_end(self):
...
-    async def __call__(self, scope, receive, send):
-        """Implements the ASGI protocol.
+    async def proxy_request(self, scope, receive, send):
+        """Wrapper for proxy request.
 
-        See details at:
-            https://asgi.readthedocs.io/en/latest/specs/index.html.
+        This method is served as common entry point by the proxy. It handles the
+        routing, including `/-/routes` and `/-/healthz`, ongoing request counter,
+        and metrics.
```
Comment (edoakes) [inline python/ray/serve/_private/http_proxy.py:418]: the paths in the docstring are http-specifc

also a lot of these methods take `scope, receive, send` which are all http-specific

Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:418]: correct, I want this first PR has no logical changes, just reorganize the methods, so we can get this merged in and I will raise the following PR to refactor using the new data models

**Comment by GeneDer** [comment]
@edoakes if there are no other comments, can we get this merge in so the following PR can be raised? I will add unit tests to both proxies when we add gRPC proxy.

**Review by edoakes — APPROVED** [review]
small nit, can be done in follow-up

File: python/ray/serve/_private/http_proxy.py [hunk 4]
```diff
@@ -574,6 +565,146 @@ async def __call__(self, scope, receive, send):
...
+    """This class is meant to be instantiated and run by an ASGI HTTP server.
+
+    >>> import uvicorn
+    >>> controller_name = ... # doctest: +SKIP
+    >>> uvicorn.run(HTTPProxy(controller_name)) # doctest: +SKIP
+    """
+
+    @property
+    def proxy_name(self) -> str:
+        return "HTTP"
```
[the remark this answers is not in the document]
Comment (GeneDer) [reply python/ray/serve/_private/http_proxy.py:657]: Haha, I was thinking the same, will address in followups!

**Pull request merged**